lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -0,0 +1,312 @@
+begin
+  require "oauth"
+rescue LoadError
+  raise "OAuth is required to use QuickBooks Payments. Please run `gem install oauth`."
+end
+
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class QuickbooksGateway < Gateway
+      self.test_url = 'https://sandbox.api.intuit.com'
+      self.live_url = 'https://api.intuit.com'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = [:visa, :master, :american_express, :discover, :diners]
+
+      self.homepage_url = 'http://payments.intuit.com'
+      self.display_name = 'QuickBooks Payments'
+      ENDPOINT =  "/quickbooks/v4/payments/charges"
+      OAUTH_ENDPOINTS = {
+        site: 'https://oauth.intuit.com',
+        request_token_path: '/oauth/v1/get_request_token',
+        authorize_url: 'https://appcenter.intuit.com/Connect/Begin',
+        access_token_path: '/oauth/v1/get_access_token'
+      }
+
+      # https://developer.intuit.com/docs/0150_payments/0300_developer_guides/error_handling
+
+      STANDARD_ERROR_CODE_MAPPING = {
+        # Fraud Warnings
+        'PMT-1000' => STANDARD_ERROR_CODE[:processing_error],   # payment was accepted, but refund was unsuccessful
+        'PMT-1001' => STANDARD_ERROR_CODE[:invalid_cvc],   # payment processed, but cvc was invalid
+        'PMT-1002' => STANDARD_ERROR_CODE[:incorrect_address],   # payment processed, incorrect address info
+        'PMT-1003' => STANDARD_ERROR_CODE[:processing_error],   # payment processed, address info couldn't be validated
+
+        # Fraud Errors
+        'PMT-2000' => STANDARD_ERROR_CODE[:incorrect_cvc],      # Incorrect CVC
+        'PMT-2001' => STANDARD_ERROR_CODE[:invalid_cvc],        # CVC check unavaliable
+        'PMT-2002' => STANDARD_ERROR_CODE[:incorrect_address],  # Incorrect address
+        'PMT-2003' => STANDARD_ERROR_CODE[:incorrect_address],  # Address info unavailable
+
+        'PMT-3000' => STANDARD_ERROR_CODE[:processing_error],   # Merchant account could not be validated
+
+        # Invalid Request
+        'PMT-4000' => STANDARD_ERROR_CODE[:processing_error],   # Object is invalid
+        'PMT-4001' => STANDARD_ERROR_CODE[:processing_error],   # Object not found
+        'PMT-4002' => STANDARD_ERROR_CODE[:processing_error],   # Object is required
+
+        # Transaction Declined
+        'PMT-5000' => STANDARD_ERROR_CODE[:card_declined],      # Request was declined
+        'PMT-5001' => STANDARD_ERROR_CODE[:card_declined],      # Merchant does not support given payment method
+
+        # System Error
+        'PMT-6000' => STANDARD_ERROR_CODE[:processing_error],   # A temporary Issue prevented this request from being processed.
+      }
+
+      FRAUD_WARNING_CODES = ['PMT-1000','PMT-1001','PMT-1002','PMT-1003']
+
+      def initialize(options = {})
+        requires!(options, :consumer_key, :consumer_secret, :access_token, :token_secret, :realm)
+        @consumer_key = options[:consumer_key]
```

Comment:
I actually prefer just accessing these values straight out of `@options` vs assigning everything to instance variables.